### DISCUSSION TITLE
API resolved without sending a response for /api/stats, this may result in stalled requests.

### DISCUSSION THREAD

**Discussion by Saggitus007** [vercel/next.js#44897] [Help]
import jwt from "jsonwebtoken";
import {
    findVideoIdByUser,
    insertStats,
    updateStats,
  } from "../../lib/db/hasura";

export default async function stats(req, resp) {
    if (req.method === "POST") {
        console.log({ cookies: req.cookies });

        try{
            const token = req.cookies.token;
            if (!token) {

                const { videoId, favourited, watched = true } = req.body;
            } else {
     
                if (videoId) {
                    const decodedToken = jwt.verify(token, process.env.NEXT_PUBLIC_JWT_SECRET );
                console.log({ decodedToken });
                
                const userId = decodedToken.issuer;
                //const videoId = "g8evyE9TuYk";
                const doesStatsExist = await findVideoIdByUser
                (token, userId, videoId);
                if (doesStatsExist) {
                    // update it
                    const response = await updateStats(token, {
                        watched,
                        userId,
                        videoId,
                        favourited,
                    });
                    resp.send({ data:
                    response });
                } else {
                    // add it
                    const response = await insertStats(token, {
                        watched,
                        userId,
                        videoId,
                        favourited,
                      });
                      resp.send({ data: response });
                    }
                }
                  }
                } catch (error) {
                  console.error("Error occurred /stats", error);
                  resp.status(500).send({ done: false, error: error?.message });
                }
              }
            }

**Comment by Saggitus007** [comment]
please l need help on this

**Comment by icyJoseph** [comment]
Hi,

Try with:

```js
import jwt from "jsonwebtoken";
import {
  findVideoIdByUser,
  insertStats,
  updateStats,
} from "../../lib/db/hasura";

export default async function stats(req, resp) {
  // exit early if the method's not POST (note that "post", lowercased, will be rejected)
  if (req.method !== "POST") return resp.status(404).end();

  try {
    const token = req.cookies.token;
    if (!token) {
      // there was no token, exit with bad request
      return resp.status(400).end();
    }
    // we did have a token - no else branch required
    const { videoId, favourited, watched = true } = req.body;

    // we do not have a videoId
    if (!videoId) return resp.status(400).end();

    const decodedToken = jwt.verify(token, process.env.NEXT_PUBLIC_JWT_SECRET);

    const userId = decodedToken.issuer;

    const doesStatsExist = await findVideoIdByUser(token, userId, videoId);

    if (doesStatsExist) {
      // update it
      const response = await updateStats(token, {
        watched,
        userId,
        videoId,
        favourited,
      });

      return resp.json({ data: response });
    }
    // since the above if-branch returns, else is not needed
    // add it
    const response = await insertStats(token, {
      watched,
      userId,
      videoId,
      favourited,
    });

    return resp.json({ data: response });
  } catch (error) {
    console.error("Error occurred /stats", error);
    return resp.status(500).send({ done: false, error: error?.message });
  }
}

```

Basically, I switched to early exits, so that it is clear when each path ends. Your original snippet contains paths that lead to no result, so the request never resolves.

I also changed `resp.send` for `resp.json`.

I did not improve the `doesStatsExist` further, but it could be also done in a shorter, clearer way. 

Let me know how it goes.

**Reply by Saggitus007** [reply]
so l get another error again from hasura thank you very much for breaking it dwon for me l am new to web development please will send the other code that leads to an error

**Comment by Saggitus007** [comment]
export async function insertStats(
  token,
  { favourited, userId, watched, videoId }
) {
  const operationsDoc = `
  mutation insertStats($favourited: Int!, $userId: String!, $watched: Boolean!, $videoId: String!) {
    insert_stats_one(object: {
      favourited: $favourited, 
      userId: $userId, 
      watched: $watched, 
      videoId: $videoId
    }) {
        favourited
        id
        userId
    }
  }
`;

  return await queryHasuraGQL(
    operationsDoc,
    "insertStats",
    { favourited, userId, watched, videoId },
    token
  );
}

export async function updateStats(
  token,
  { favourited, userId, watched, videoId }
) {
  const operationsDoc = `
mutation updateStats($favourited: Int!, $userId: String!, $watched: Boolean!, $videoId: String!) {
  update_stats(
    _set: {watched: $watched, favourited: $favourited}, 
    where: {
      userId: {_eq: $userId}, 
      videoId: {_eq: $videoId}
    }) {
    returnig {
      favourited,
      userId,
      watched,
      videoId
    }
  }
}
`;
  return await queryHasuraGQL(
    operationsDoc,
    "updateStats",
    { favourited, userId, watched, videoId },
    token
  );
}
export async function findVideoIdByUser(token, userId, videoId) {
  const operationsDoc = `
  query findVideoIdByUserId($userId: String!, $videoId: String!) {
    stats(where: { userId: {_eq: $userId}, videoId: {_eq: $videoId }}) {
      id
      userId
      videoId
      favourited
      watched
    }
  }
`;

const response = await queryHasuraGQL(
  operationsDoc,
  "findVideoIdByUserId",
  {
    videoId,
    userId,
  },
  token
);

return response?.data?.stats;
}


export async function createNewUser(token, metadata) {
  const operationsDoc = `
  mutation createNewUser($issuer: String!, $email: String!, $publicAddress: String!) {
    insert_users(objects: {email: $email, issuer: $issuer, publicAddress: $publicAddress}) {
      returning {
        email
        id
        issuer
      }
    }
  }
`;

  const { issuer, email, publicAddress } = metadata;
  const response = await queryHasuraGQL(
    operationsDoc,
    "createNewUser",
    {
      issuer,
      email,
      publicAddress,
    },
    token
  );
  console.log({ response, issuer });
  return response;
}

export async function isNewUser(token, issuer) {
  const operationsDoc = `
  query isNewUser($issuer: String!) {
    users(where: {issuer: {_eq: $issuer}}) {
      id
      email
      issuer
    }
  }
`;
  const response = await queryHasuraGQL(
    operationsDoc,
    "isNewUser",
    {
      issuer,
    },
    token
  );
  console.log({ response, issuer });
  return response?.data?.users?.length === 0;
}
async function queryHasuraGQL(operationsDoc, operationName, variables, token) {
  const result = await fetch(process.env.NEXT_PUBLIC_HASURA_ADMIN_URL, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-type": "application/json",
    },
    body: JSON.stringify({
      query: operationsDoc,
      variables: variables,
      operationName: operationName,
    }),
  });
  return await result.json();
}

**Comment by Saggitus007** [comment]
so this is the error 

data 
{data: {…}}
data
: 
errors
: 
Array(1)
0
: 
{extensions: {…}, message: "field 'videoId' not found in type: 'stats_insert_input'"}
length
: 
1
[[Prototype]]
: 
Array(0)
[[Prototype]]
: 
Object
[[Prototype]]
: 
Object
